cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,309 @@
+import co.cask.tephra.runtime.TransactionModules;
+import com.google.common.base.Function;
+import com.google.common.collect.Multimap;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.gson.Gson;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Singleton;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NavigableMap;
+import javax.annotation.Nullable;
+
+/**
+ * Debugging tool for queues in hbase.
+ */
+public class HBaseQueueDebugger extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HBaseQueueDebugger.class);
+  private static final Gson GSON = new Gson();
+
+  private final HBaseQueueAdmin queueAdmin;
+  private final TransactionServiceClient txClient;
+  private final ZKClientService zkClientService;
+  private final HBaseQueueClientFactory queueClientFactory;
+
+  @Inject
+  public HBaseQueueDebugger(HBaseQueueAdmin queueAdmin,
+                            HBaseQueueClientFactory queueClientFactory,
+                            TransactionServiceClient txClient,
+                            ZKClientService zkClientService) {
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.txClient = txClient;
+    this.zkClientService = zkClientService;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    zkClientService.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    zkClientService.stopAndWait();
+  }
+
+  /**
+   * Only works for {@link co.cask.cdap.data2.transaction.queue.hbase.ShardedHBaseQueueStrategy}.
+   */
+  private void scanQueue(URI queueNameURI) throws Exception {
+    QueueName queueName = QueueName.from(queueNameURI);
+    HBaseConsumerStateStore stateStore = queueAdmin.getConsumerStateStore(queueName);
+
+    Transaction tx = txClient.startLong();
+    stateStore.startTx(tx);
+
+    Multimap<Long, QueueBarrier> barriers = stateStore.getAllBarriers();
+    LOG.info("Got {} barriers", barriers.size());
+
+    QueueStatistics stats = new QueueStatistics();
+    int currentSection = 1;
+    for (Long groupId : barriers.keySet()) {
+      LOG.info("Scanning barriers for group {}", groupId);
+      Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+      Iterator<QueueBarrier> barrierIterator = groupBarriers.iterator();
+
+      QueueBarrier previous;
+      QueueBarrier current = null;
+      while (barrierIterator.hasNext()) {
```

Comment:
You don't need order by key (group ID), you just need to maintain the ordering of the barriers with the same order of insertion (scanning from HBase gives you from earliest to latest barrier).